File: frontend/src/app/core/services/cliente.service.ts
import { HttpClient, HttpParams } from '@angular/common/http';
import { Injectable, inject } from '@angular/core';
import { Observable } from 'rxjs';
import { Cliente, ClienteRequest, InteressePrincipal } from '../models/cliente.model';
import { Page } from '../models/page.model';
import { RuntimeConfigService } from './runtime-config.service';

export interface ClienteFiltro {
  interesse?: InteressePrincipal | '';
  q?: string;
  page?: number;
  size?: number;
}

@Injectable({ providedIn: 'root' })
export class ClienteService {
  private readonly http = inject(HttpClient);
  private readonly config = inject(RuntimeConfigService);

  private get baseUrl(): string {
    return `${this.config.apiUrl}/api/clientes`;
  }

  listar(filtro: ClienteFiltro): Observable<Page<Cliente>> {
    let params = new HttpParams()
      .set('page', filtro.page ?? 0)
      .set('size', filtro.size ?? 10);
    if (filtro.interesse) params = params.set('interesse', filtro.interesse);
    if (filtro.q) params = params.set('q', filtro.q);
    return this.http.get<Page<Cliente>>(this.baseUrl, { params });
  }

  buscarPorId(id: number): Observable<Cliente> {
    return this.http.get<Cliente>(`${this.baseUrl}/${id}`);
  }

  criar(request: ClienteRequest): Observable<Cliente> {
    return this.http.post<Cliente>(this.baseUrl, request);
  }

  atualizar(id: number, request: ClienteRequest): Observable<Cliente> {
    return this.http.put<Cliente>(`${this.baseUrl}/${id}`, request);
  }

  excluir(id: number): Observable<void> {
    return this.http.delete<void>(`${this.baseUrl}/${id}`);
  }
}

File: backend/src/main/java/com/impar/crmcarros/cliente/ClienteService.java
package com.impar.crmcarros.cliente;

import com.impar.crmcarros.cliente.dto.ClienteRequest;
import com.impar.crmcarros.cliente.dto.ClienteResponse;
import com.impar.crmcarros.common.exception.BusinessRuleException;
import com.impar.crmcarros.common.exception.ResourceNotFoundException;
import lombok.RequiredArgsConstructor;
import org.springframework.dao.DataIntegrityViolationException;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.data.jpa.domain.Specification;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.ArrayList;
import java.util.List;

@Service
@RequiredArgsConstructor
@Transactional
public class ClienteService {

    private final ClienteRepository repository;

    @Transactional(readOnly = true)
    public Page<ClienteResponse> listar(InteressePrincipal interesse, String q, Pageable pageable) {
        List<Specification<Cliente>> specs = new ArrayList<>();
        if (interesse != null) {
            specs.add(ClienteSpecifications.comInteresse(interesse));
        }
        if (q != null && !q.isBlank()) {
            specs.add(ClienteSpecifications.comTexto(q.trim()));
        }
        Specification<Cliente> spec = specs.stream().reduce(Specification::and).orElse(null);
        return repository.findAll(spec, pageable).map(ClienteMapper::toResponse);
    }

    @Transactional(readOnly = true)
    public ClienteResponse buscarPorId(Long id) {
        return ClienteMapper.toResponse(buscarEntidadePorId(id));
    }

    public ClienteResponse criar(ClienteRequest request) {
        validarEmailDisponivel(request.email(), null);
        Cliente cliente = salvarTratandoEmailDuplicado(ClienteMapper.toEntity(request));
        return ClienteMapper.toResponse(cliente);
    }

    public ClienteResponse atualizar(Long id, ClienteRequest request) {
        Cliente cliente = buscarEntidadePorId(id);
        validarEmailDisponivel(request.email(), id);
        ClienteMapper.updateEntity(cliente, request);
        return ClienteMapper.toResponse(salvarTratandoEmailDuplicado(cliente));
    }

    /**
     * A checagem de {@link #validarEmailDisponivel} evita a maioria dos casos, mas nao
     * fecha uma corrida de concorrencia entre duas requisicoes simultaneas com o mesmo
     * e-mail; a constraint UNIQUE do banco e a rede de seguranca final para esse caso.
     */
    private Cliente salvarTratandoEmailDuplicado(Cliente cliente) {
        try {
            return repository.saveAndFlush(cliente);
        } catch (DataIntegrityViolationException ex) {
            throw new BusinessRuleException("Ja existe um cliente cadastrado com o e-mail " + cliente.getEmail());
        }
    }

    public void excluir(Long id) {
        Cliente cliente = buscarEntidadePorId(id);
        try {
            repository.delete(cliente);
            repository.flush();
        } catch (DataIntegrityViolationException ex) {
            throw new BusinessRuleException(
                    "Nao e possivel excluir este cliente pois ele possui oportunidades de venda associadas.");
        }
    }

    private void validarEmailDisponivel(String email, Long idAtual) {
        repository.findByEmailIgnoreCase(email)
                .filter(c -> idAtual == null || !c.getId().equals(idAtual))
                .ifPresent(c -> {
                    throw new BusinessRuleException("Ja existe um cliente cadastrado com o e-mail " + email);
                });
    }

    private Cliente buscarEntidadePorId(Long id) {
        return repository.findById(id)
                .orElseThrow(() -> ResourceNotFoundException.of("Cliente", id));
    }
}
